dpc/src/account/private_key.rs

```diff
@@ -119,90 +117,92 @@ impl<C: Parameters> PrivateKey<C> {
-        let decryption_key = match self.is_dummy {
-            true => <C::AccountEncryptionScheme as EncryptionScheme>::PrivateKey::default(),
-            false => <C::AccountEncryptionScheme as EncryptionScheme>::PrivateKey::read_le(&decryption_key_bytes[..])?,
+        // Compute the commitment, which is used as the decryption key.
+        let commitment = C::account_commitment_scheme().commit(&self.commitment_input, &self.r_pk)?;
+        let commitment_bytes = commitment.to_bytes_le()?;
+
+        // Determine the number of MSB bits we must enforce are zero,
+        // for the isomorphism from the base field to the scalar field.
+        let enforce_zero_on_num_bits = {
+            debug_assert!(C::AccountEncryptionScheme::private_key_size_in_bits() > 0);
+            // We must enforce that the MSB bit of the scalar field is also set to 0.
+            let capacity = C::AccountEncryptionScheme::private_key_size_in_bits() - 1;
+            let commitment_num_bits = commitment_bytes.len() * 8;
+            assert!(capacity < commitment_num_bits);
+
+            commitment_num_bits - capacity
         };
 
         // This operation explicitly enforces that the unused MSB bits
         // for the scalar field representation are correctly set to 0.
-        //
-        // To simplify verification of this isomorphism from the base field
-        // to the scalar field in the `inner_snark`, we additionally enforce
-        // that the MSB bit of the scalar field is also set to 0.
-        if !self.is_dummy {
-            let account_decryption_key_bits = from_bytes_le_to_bits_le(&decryption_key_bytes[..]).collect::<Vec<_>>();
-            let account_decryption_key_length = account_decryption_key_bits.len();
-
-            let decryption_private_key_length = C::AccountEncryptionScheme::private_key_size_in_bits();
-            assert!(decryption_private_key_length > 0);
-            assert!(decryption_private_key_length <= account_decryption_key_length);
-
-            for i in (decryption_private_key_length - 1)..account_decryption_key_length {
-                let bit_index = account_decryption_key_length - i - 1;
-                if account_decryption_key_bits[bit_index] {
-                    return Err(AccountError::InvalidAccountCommitment);
-                }
+        for msb_bit in from_bytes_le_to_bits_le(&commitment_bytes[..])
+            .rev()
+            .take(enforce_zero_on_num_bits)
+        {
+            // Pop the next MSB bit, and enforce it is zero.
+            if msb_bit {
+                return Err(AccountError::InvalidAccountCommitment);
             }
         }
 
+        // This operation implicitly enforces that the unused MSB bits
+        // for the scalar field representation are correctly set to 0.
+        let decryption_key: <C::AccountEncryptionScheme as EncryptionScheme>::PrivateKey =
+            FromBytes::read_le(&commitment_bytes[..])?;
+
```

Comment:
We may want to qualify this "implicit check" more. It is weaker than the MSB check:
- In MSB check, we ensure that it is within the capacity of the TE scalar field.
- In this implicit check (due to ToBytes), it ensures that the number is smaller than the TE curve's scalar field prime number, which may still be larger than the capacity.

Reply:
Do you mean, add a comment that states this explicitly, or should there be something stricter being done?

I will add the comment right now.

Reply:
I went ahead and added this check into the function as well:
```rs
        // Enforce the MSB of the scalar field element is 0 by convention.
        debug_assert_eq!(Some(&false), decryption_key.to_bits_be().iter().next());
```

For now, I will leave it as a `debug_assert_eq` so it does not affect production performance. However, if we feel the need to elevate it to a release assertion, feel free to make the change.